lib/checkassert.cpp

```diff
@@ -53,45 +54,103 @@ void CheckAssert::assertWithSideEffects()
-
-                bool noReturnInScope = true;
-                for (const Token *rt = scope->bodyStart; rt != scope->bodyEnd; rt = rt->next()) {
-                    if (rt->str() != "return") continue; // find all return statements
-                    if (inSameScope(rt, tok2)) {
-                        noReturnInScope = false;
-                        break;
+        }
+        tok = endTok;
+    }
+}
+
+bool CheckAssert::isFunctionWithSideEffect(const Function *function, argumentCheck argsChecking) const {
+    if (function->isConstexpr()) return false;
+    if (function->nestedIn->isClassOrStruct() && !function->isConstructor()) {
+        if (!(function->isConst() || function->isStatic())) {
+            return true;
+        }
+    }
+    const Scope *scope = function->functionScope;
+    if (!scope) return false;
+
+    if (function->isConstructor()) {
+        const auto *initializationList =
+            function->constructorMemberInitialization();
+        if (initializationList) {
+            for (const Token *tok = initializationList; tok != scope->bodyStart; tok = tok->next()) {
+                if (tok->tokType() == Token::eIncDecOp) {
+                    const Variable *var = tok->previous()->variable();
+                    if (!var)
+                        continue;
+                    if (var->isArgument() && argsChecking.arguments) {
+                        return checkArgument(tok, function, var, argsChecking);
                     }
                 }
-                if (noReturnInScope) continue;
-
-                sideEffectInAssertError(tmp, f->name());
-                break;
             }
         }
-        tok = endTok;
     }
+
+    for (const Token *tok = scope->bodyStart; tok != scope->bodyEnd; tok = tok->next()) {
+        if (tok->tokType() == Token::eFunction) {
+            // previously function calls on a second level from assert aren't checked at all
+            // now check at least if static or global variables are changed
+            if (isFunctionWithSideEffect(tok->function()))
+                return true;
+            continue;
+        }
+        if (!tok->isAssignmentOp() && tok->tokType() != Token::eIncDecOp)
+            continue;
+
+        const Variable *var = tok->previous()->variable();
+        if (!var)
+            continue;
+
+        if (var->isLocal())
+            continue;
+        if (var->isStatic() || var->isGlobal())
+            return true;
+
+        if (var->isArgument() && argsChecking.arguments) {
+            return checkArgument(tok, function, var, argsChecking);
+        }
+    }
+    return false;
+}
+
+const Variable *CheckAssert::findPassedVariable(const Function *function, const Variable *var, const argumentCheck &argsChecking) {
+    int paramCtr = 0;
+    const Variable *arg;
+    const Variable *variable = nullptr;
+    do {
+        arg = function->getArgumentVar(paramCtr);
+        if (arg == var) {
```

Comment:
looks like a infinite loop if `arg != var`. I suggest a for loop instead of do while.